mlfinlab/online_portfolio_selection/online_portfolio_selection.py

```diff
@@ -0,0 +1,330 @@
+
+        :param optimize_array: (np.array) Relative returns of the assets for a given time period.
+        :param solver: (cp.solver) Solver for cvxpy
+        :return weights.value: (np.array) Weights that maximize the returns for the given array.
+        """
+
+        # Initialize weights for the optimization problem.
+        weights = cp.Variable(self.number_of_assets)
+
+        # Use cp.log and cp.sum to make the cost function a convex function.
+        # Multiplying continuous returns equates to summing over the log returns.
+        portfolio_return = cp.sum(cp.log(optimize_array * weights))
+
+        # Optimization objective and constraints.
+        allocation_objective = cp.Maximize(portfolio_return)
+        allocation_constraints = [cp.sum(weights) == 1, cp.min(weights) >= 0]
+
+        # Define and solve the problem.
+        problem = cp.Problem(objective=allocation_objective, constraints=allocation_constraints)
+
+        # Solve and return the resulting weights.
+        problem.solve(warm_start=True, solver=solver)
+        return weights.value
+
+    def _round_weights(self, threshold=1e-6):
+        """
+        Drops weights that are below a certain threshold.
+
+        :param threshold: (float) Drop all values below this threshold.
+        """
+        # Set all values below the threshold to 0.
+        new_all_weights = np.where(self.all_weights < threshold, 0, self.all_weights)
+
+        # Adjust weights to have the weights sum to 1.
+        new_all_weights = np.apply_along_axis(lambda x: x / np.sum(x), 1, new_all_weights)
+        self.all_weights = new_all_weights
+
+    def _uniform_weight(self):
+        """
+        Returns a uniform weight of assets.
+
+        :return uni_weight: (np.array) Uniform weights (1/n, 1/n, 1/n ...).
+        """
+        # Divide by number of assets after creating numpy arrays of one.
+        uni_weight = np.ones(self.number_of_assets) / self.number_of_assets
+        return uni_weight
+
+    def _print_progress(self, iteration, prefix='', suffix='', decimals=1, bar_length=50):
+        # pylint: disable=expression-not-assigned
+        """
+        Calls in a loop to create a terminal progress bar.
+        https://gist.github.com/aubricus/f91fb55dc6ba5557fbab06119420dd6a
+
+        :param iteration: (int) Current iteration.
+        :param prefix: (str) Prefix string.
+        :param suffix: (str) Suffix string.
+        :param decimals: (int) Positive number of decimals in percent completed.
+        :param bar_length: (int) Character length of the bar.
+        """
+        str_format = "{0:." + str(decimals) + "f}"
+        # Calculate the percent completed.
+        percents = str_format.format(100 * (iteration / float(self.length_of_time)))
+        # Calculate the length of bar.
+        filled_length = int(round(bar_length * iteration / float(self.length_of_time)))
+        # Fill the bar.
+        block = '█' * filled_length + '-' * (bar_length - filled_length)
```

Comment:
What is this black character block? Is it intended?

Reply:
If you run the method, the blocks fill up the progress bar.

<img width="676" alt="Screen Shot 2020-04-27 at 2 48 45 PM" src="https://user-images.githubusercontent.com/29555998/80424255-32faf180-8896-11ea-9575-6d8bcf197b30.png">